### PR TITLE
Enable global planner tester

### PR DESCRIPTION
**Pull request by orduno**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #586 |
| Primary OS tested on | Ubuntu 18.0 |
| Robotic platform tested on | N/A |

---

## Description of contribution in a few bullet points

* Made a few upgrades, including using the ROS2 actions interface.
* Enabling the planner tests on CI. Ran the tests on my machine 100 times with two failures -- planning service was not available on both.

### CODE REVIEW

File: nav2_system_tests/src/planning/planner_tester.cpp [hunk 1]
```diff
@@ -440,7 +491,7 @@ bool PlannerTester::sendCancel()
   return false;
 }
 
-void PlannerTester::printPath(const nav2_behavior_tree::ComputePathToPoseResult & path) const
+void PlannerTester::printPath(const ComputePathToPoseResult & path) const
```
Comment (bpwilcox) [inline nav2_system_tests/src/planning/planner_tester.cpp]: you could consider using a `std::stringstream` instead of printing multiple `RCLCPP_INFO` lines

**Comment by mkhansen-intel** [comment]
Can you rebase and resolve the conflict shown in planner_tester.cpp, so we can hopefully get this passing in CI?

**Comment by mkhansen-intel** [comment]
--- stderr: nav2_system_tests
In file included from /opt/nav2_ws/src/navigation2/nav2_system_tests/src/planning/test_planner_node.cpp:20:0:
/opt/nav2_ws/src/navigation2/nav2_system_tests/src/planning/planner_tester.hpp:29:10: fatal error: nav2_msgs/srv/get_robot_pose.hpp: No such file or directory
 #include "nav2_msgs/srv/get_robot_pose.hpp"
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.

File: nav2_system_tests/src/planning/planner_tester.cpp [hunk 2]
```diff
@@ -243,31 +266,36 @@ bool PlannerTester::defaultPlannerRandomTests(
...
   auto generate_random = [&]() mutable -> std::pair<int, int> {
       bool point_is_free = false;
       int x, y;
       while (!point_is_free) {
         x = distribution_x(generator);
         y = distribution_y(generator);
-        point_is_free = costmap_->isFree(x, y);
+        point_is_free = costmap_->is_free(x, y);
       }
       return std::make_pair(x, y);
```
[the remark this answers is not in the document]
Comment (orduno) [reply nav2_system_tests/src/planning/planner_tester.cpp]: For this PR, prefer not to change the interface of other packages.

**Comment by orduno** [comment]
Since the robot pose is now provided to the planner through a transform, I am working on updating the planner tester to provide this transform.

**Comment by mkhansen-intel** [comment]
It's still failing CI:
```
[navfn_planner-1] [INFO] [navfn_planner_GetCostmap_client_42307709]: GetCostmap service client: waiting for service to appear...
```

**Comment by orduno** [comment]
Yeah, I wasn't finished with the changes, but it should be working now :smiley: let's see what CI thinks...

Ran the test on my machine 100 times (each test sends 100 random plan requests). Failed 3 tests out of 100, but the fails were related to FastRTPS:

```bash
[navfn_planner-1] [INFO] [navfn_planner]: Activating
[lifecycle_manager-2] [INFO] [lifecycle_manager]: The system is active
[navfn_planner-1] navfn_planner: /home/caorduno/master/ros2_ws/src/eProsima/Fast-RTPS/src/cpp/rtps/reader/WriterProxy.cpp:338: void eprosima::fastrtps::rtps::WriterProxy::change_removed_from_history(const eprosima::fastrtps::rtps::SequenceNumber_t&): Assertion `chit != changes_received_.begin()' failed.
```

File: nav2_system_tests/src/planning/planner_tester.cpp [hunk 3]
```diff
@@ -319,47 +339,67 @@ bool PlannerTester::plannerTest(
...
-  // Loop until the subtask is completed
-  while (true) {
-    TaskStatus status = planner_client_->waitForResult(path, std::chrono::milliseconds(100));
+  auto future_result = planner_client_->async_get_result(goal_handle);
 
-    if (status != TaskStatus::RUNNING) {
-      return status;
-    }
+  RCLCPP_DEBUG(this->get_logger(), "Wait for the server to be done with the action");
+  auto status_result = future_result.wait_for(seconds(1));
```
Comment (mkhansen-intel) [inline nav2_system_tests/src/planning/planner_tester.cpp]: Is 1 second long enough to wait for the result?

File: nav2_system_tests/src/planning/planner_tester.cpp [hunk 4]
```diff
@@ -440,14 +480,28 @@ bool PlannerTester::sendCancel()
...
   }
+
+  RCLCPP_INFO(get_logger(), ss.str().c_str());
 }
 
+void PlannerTester::waitForPlanner()
+{
+  RCLCPP_DEBUG(this->get_logger(), "Waiting for ComputePathToPose action server");
+
+  if (!planner_client_ || !planner_client_->wait_for_action_server(4s)) {
```
Comment (mkhansen-intel) [inline nav2_system_tests/src/planning/planner_tester.cpp]: Is 4 seconds always enough time to wait for the action server to be ready?

**Comment by mkhansen-intel** [comment]
@orduno - I just looked at the output and it seems the test returns 'PASS' even when the planner failed to find a path 39/100 times. Is this expected?
```
9: [navfn_planner-1] [INFO] [navfn_planner]: Preempting the goal pose.
9: [test_planner_node-3] [WARN] [PlannerTester]: Failed with start at 175.00, 41.00 and goal at 253.00, 392.00
9: [test_planner_node-3] [WARN] [PlannerTester]: Failed with start at 245.00, 270.00 and goal at 32.00, 102.00
9: [test_planner_node-3] [WARN] [PlannerTester]: Failed with start at 166.00, 203.00 and goal at 135.00, 371.00
9: [navfn_planner-1] [INFO] [navfn_planner]: Preempting the goal pose.
9: [test_planner_node-3] [WARN] [PlannerTester]: Failed with start at 80.00, 99.00 and goal at 45.00, 24.00
9: [test_planner_node-3] [WARN] [PlannerTester]: Failed with start at 250.00, 314.00 and goal at 335.00, 110.00
9: [navfn_planner-1] [INFO] [navfn_planner]: Preempting the goal pose.
9: [test_planner_node-3] [INFO] [PlannerTester]: Tested with 100 tests. Planner failed on 39. Test time 54045 ms
9: [test_planner_node-3] [       OK ] PlannerTester.testWithHundredRandomEndPoints (54135 ms)
9: [test_planner_node-3] [----------] 2 tests from PlannerTester (54245 ms total)
9: [test_planner_node-3] 
9: [test_planner_node-3] [----------] Global test environment tear-down
9: [test_planner_node-3] [==========] 2 tests from 1 test case ran. (54249 ms total)
9: [test_planner_node-3] [  PASSED  ] 2 tests.
```

**Pull request merged**